core/src/main/java/org/infinispan/interceptors/locking/AbstractLockingInterceptor.java

```diff
@@ -103,12 +103,16 @@ public Object visitGetCacheEntryCommand(InvocationContext ctx, GetCacheEntryComm
 
    // We need this method in here because of putForExternalRead
    protected final Object visitNonTxDataWriteCommand(InvocationContext ctx, DataWriteCommand command) throws Throwable {
-      if (hasSkipLocking(command) || !shouldLockKey(command.getKey())) {
+      Object key;
```

Comment:
Have you handled this already? It's good that you've added the tests, but as I've mentioned below these seem to me are flawed.